java/vector/src/main/java/org/apache/arrow/vector/file/ArrowWriter.java

```diff
@@ -17,81 +17,242 @@
-    checkStarted();
-    ArrowBlock batchDesc = MessageSerializer.serialize(out, recordBatch);
-    LOGGER.debug(String.format("RecordBatch at %d, metadata: %d, body: %d",
-        batchDesc.getOffset(), batchDesc.getMetadataLength(), batchDesc.getBodyLength()));
-
-    // add metadata to footer
-    recordBatches.add(batchDesc);
-  }
-
-  private void checkStarted() throws IOException {
-    if (!started) {
-      started = true;
-      start();
-    }
-  }
-
-  @Override
-  public void close() throws IOException {
-    try {
-      long footerStart = out.getCurrentPosition();
-      writeFooter();
-      int footerLength = (int)(out.getCurrentPosition() - footerStart);
-      if (footerLength <= 0 ) {
-        throw new InvalidArrowFileException("invalid footer");
-      }
-      out.writeIntLittleEndian(footerLength);
-      LOGGER.debug(String.format("Footer starts at %d, length: %d", footerStart, footerLength));
-      writeMagic();
-    } finally {
-      out.close();
-    }
-  }
-
-  private void writeMagic() throws IOException {
-    out.write(ArrowReader.MAGIC);
-    LOGGER.debug(String.format("magic written, now at %d", out.getCurrentPosition()));
-  }
-
-  private void writeFooter() throws IOException {
-    // TODO: dictionaries
-    out.write(new ArrowFooter(schema, Collections.<ArrowBlock>emptyList(), recordBatches), false);
-  }
+    private boolean allocated = false;
+
+    /**
+     * Note: fields are not closed when the writer is closed
+     *
+     * @param schema
+     * @param out
+     * @param allocator
+     */
+    protected ArrowWriter(Schema schema, OutputStream out, BufferAllocator allocator) {
+        this(schema.getFields(), createVectors(schema.getFields(), allocator), Channels.newChannel(out), true);
+    }
+
+    protected ArrowWriter(Schema schema, WritableByteChannel out, BufferAllocator allocator) {
+        this(schema.getFields(), createVectors(schema.getFields(), allocator), out, true);
+    }
+
+    protected ArrowWriter(List<Field> fields, List<FieldVector> vectors, OutputStream out) {
+        this(fields, vectors, Channels.newChannel(out), false);
+    }
+
+    protected ArrowWriter(List<Field> fields, List<FieldVector> vectors, WritableByteChannel out, boolean allocated) {
+        this.vectors = vectors;
+        this.out = new WriteChannel(out);
+        this.allocated = allocated;
+
+        // translate dictionary fields from in-memory format to message format
+        // add dictionary ids, change field types to dictionary type instead of index type
+        List<Field> updatedFields = new ArrayList<>(fields);
+        List<ArrowDictionaryBatch> dictionaryBatches = new ArrayList<>();
+        Set<Long> dictionaryIds = new HashSet<>();
+
+        // go through to add dictionary id to the schema fields and to unload the dictionary batches
+        for (FieldVector vector: vectors) {
+            if (vector instanceof DictionaryVector) {
```

Comment:
I guess I need to examine the minor type stuff more closely so I understand this. We don't have the concept of a minor type in the C++, because the data type class is the logical type. Seems like things are inverted here (the type classes are physical types, and the minor type is the logical type), and that is making things more difficult. I'm sorry if I've created more work with incorrect cross-communication